xds/internal/resolver/cluster_specifier_plugin_test.go

```diff
@@ -0,0 +1,368 @@
+        "cluster_specifier_plugin:cspA":{
+          "childPolicy":[{"csp_experimental":{"arbitrary_field":"anythingA"}}]
+        }
+      }
+    }}]}`
+
+	wantSCParsed := internal.ParseServiceConfigForTesting.(func(string) *serviceconfig.ParseResult)(wantJSON)
+	if !internal.EqualServiceConfigForTesting(rState.ServiceConfig.Config, wantSCParsed.Config) {
+		t.Errorf("ClientConn.UpdateState received different service config")
+		t.Error("got: ", cmp.Diff(nil, rState.ServiceConfig.Config))
+		t.Fatal("want: ", cmp.Diff(nil, wantSCParsed.Config))
+	}
+
+	cs := iresolver.GetConfigSelector(rState)
+	if cs == nil {
+		t.Fatal("received nil config selector")
+	}
+
+	res, err := cs.SelectConfig(iresolver.RPCInfo{Context: context.Background()})
+	if err != nil {
+		t.Fatalf("Unexpected error from cs.SelectConfig(_): %v", err)
+	}
+
+	cluster := clustermanager.GetPickedClusterForTesting(res.Context)
+	clusterWant := clusterSpecifierPluginPrefix + "cspA"
+	if cluster != clusterWant {
+		t.Fatalf("cluster: %+v, want: %+v", cluster, clusterWant)
+	}
+	// delay res.OnCommitted()
+
+	// Perform TWO updates to ensure the old config selector does not hold a reference to cspA
+	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
+		VirtualHosts: []*xdsresource.VirtualHost{
+			{
+				Domains: []string{targetStr},
+				Routes:  []*xdsresource.Route{{Prefix: newStringP(""), ClusterSpecifierPlugin: "cspB"}},
+			},
+		},
+		// Top level csp config here - the value of cspB should get directly
+		// placed as a child policy of xds cluster manager.
+		ClusterSpecifierPlugins: map[string]clusterspecifier.BalancerConfig{"cspB": []map[string]interface{}{{"csp_experimental": cspConfig{ArbitraryField: "anythingB"}}}},
+	}, nil)
+	tcc.stateCh.Receive(ctx) // Ignore the first update.
+
+	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
+		VirtualHosts: []*xdsresource.VirtualHost{
+			{
+				Domains: []string{targetStr},
+				Routes:  []*xdsresource.Route{{Prefix: newStringP(""), ClusterSpecifierPlugin: "cspB"}},
+			},
+		},
+		// Top level csp config here - the value of cspB should get directly
+		// placed as a child policy of xds cluster manager.
+		ClusterSpecifierPlugins: map[string]clusterspecifier.BalancerConfig{"cspB": []map[string]interface{}{{"csp_experimental": cspConfig{ArbitraryField: "anythingB"}}}},
+	}, nil)
+
+	gotState, err = tcc.stateCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Error waiting for UpdateState to be called: %v", err)
+	}
+	rState = gotState.(resolver.State)
+	if err := rState.ServiceConfig.Err; err != nil {
+		t.Fatalf("ClientConn.UpdateState received error in service config: %v", rState.ServiceConfig.Err)
+	}
+	wantJSON2 := `{"loadBalancingConfig":[{
+    "xds_cluster_manager_experimental":{
+      "children":{
+        "cluster_specifier_plugin:cspA":{
+          "childPolicy":[{"csp_experimental":{"arbitrary_field":"anythingA"}}]
+        },
+		"cluster_specifier_plugin:cspB":{
```

Comment:
New comment from old comment thread, since the code was moved:

This should line up with "cluster_specifier_plugin:cspA" (outdent 1 tab).  And the same situation below @ line 356.

It's not a correctness thing, it's just formatting.